biz.aQute.bndlib/src/aQute/bnd/plugin/jpms/JPMSModuleInfoPlugin.java

```diff
@@ -246,87 +286,105 @@ private void openPackages(Analyzer analyzer, ModuleInfoBuilder builder) {
 
 				if (attrs == null || !attrs.containsKey(INTERNAL_MODULE_DIRECTIVE)) {
 					String eeModuleName = moduleEE.getModules()
 						.entrySet()
 						.stream()
 						.filter(it -> it.getValue()
-							.getTyped(Attrs.LIST_STRING, "exports")
+							.getTyped(Attrs.LIST_STRING, EXPORTS_ATTRIBUTE)
 							.contains(packageRef.getFQN()))
 						.map(Entry::getKey)
 						.findAny()
 						.orElse(null);
 
-					if (eeModuleName != null) {
-						index.put(packageRef, Attrs.create(INTERNAL_MODULE_DIRECTIVE, eeModuleName));
-						return true;
-					}
+					if (eeModuleName == null) {
+						if (logger.isWarnEnabled())
+							logger.warn("Can't find a module name for imported package: {}", packageRef.getFQN());
 
-					if (logger.isWarnEnabled())
-						logger.warn("Can't find a module name for imported package: {}", packageRef.getFQN());
+						return false;
+					}
 
-					return false;
+					attrs = Attrs.create(INTERNAL_MODULE_DIRECTIVE, eeModuleName);
+					index.put(packageRef, attrs);
+				}
+				if (importAttrs != null) {
+					attrs.mergeWith(importAttrs, false);
 				}
 				return true;
 			})
 			// group packages by module/contract
 			.collect(groupingBy(entry -> index.get(entry.getKey())
-				.get(INTERNAL_MODULE_DIRECTIVE)))
-			.entrySet()
+				.get(INTERNAL_MODULE_DIRECTIVE)));
+
+		String manuallyRequiredModules = instruction.getValue()
+			.get(MODULES_ATTRIBUTE);
+
+		if (manuallyRequiredModules != null) {
+			splitAsStream(manuallyRequiredModules)
+				.forEach(moduleToAdd -> {
+					requiresMap.computeIfAbsent(moduleToAdd, key -> emptyList());
+				});
+		}
+
+		requiresMap.entrySet()
+			.stream()
+			.sorted(comparingByKey(naturalOrder()))
 			.forEach(entry -> {
 				String moduleName = entry.getKey();
 
-				if (requiredModules.contains(moduleName)) {
+				List<Entry<PackageRef, Attrs>> referencedPackages = entry.getValue();
+
+				List<PackageRef> referencedModulePackages = referencedPackages.stream()
+					.map(Entry::getKey)
+					.collect(toList());
+
+				Attrs moduleMappingAttrs = Optional.ofNullable(moduleInfoOptions.get(moduleName))
+					.orElse(new Attrs());
+
+				if (isTrue(moduleMappingAttrs.get(IGNORE_ATTRIBUTE))) {
 					return;
 				}
 
-				List<Entry<PackageRef, Attrs>> referencedPackages = entry.getValue();
-
 				// An import results in `transitive` requires where there is an
 				// `Export-Package` that has a `uses` constraint on the imported
 				// package.
-				boolean isTransitive = exports.entrySet()
-					.stream()
-					.map(export -> export.getValue()
-						.get(USES_DIRECTIVE))
-					.flatMap(Strings::splitAsStream)
-					.map(use -> analyzer.getPackageRef(Descriptors.fqnToBinary(use)))
-					.anyMatch(packageRef -> referencedPackages.stream()
-						.anyMatch(e -> e.getKey()
-							.equals(packageRef)));
+				boolean isTransitive = Optional.ofNullable(moduleMappingAttrs.get(TRANSITIVE_ATTRIBUTE))
+					.map(Processor::isTrue)
+					.orElse(exports.entrySet()
```

Comment:
You should make this `orElseGet` to avoid all the work until it is needed. `orElse` is good when you already have the else value handy. But it you have to go to some work to make the else value, `orElseGet` can avoid that work when you don't need the else value.